File: scala-server/app/com/bolour/boardgame/scala/server/domain/GameBase.scala
/*
 * Copyright 2017-2018 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
 */
package com.bolour.boardgame.scala.server.domain

import java.time.Instant

import com.bolour.boardgame.scala.common.domain.PieceProviderType.PieceProviderType
import com.bolour.util.scala.common.CommonUtil.ID
import com.bolour.util.scala.server.BasicServerUtil.stringId
import com.bolour.boardgame.scala.common.domain.{GameParams, InitPieces, Piece, PiecePoint}
import com.bolour.boardgame.scala.server.util.WordUtil.english

/**
  * Properties of a game that are conceptually immutable once set.
  *
  * @param pointValues Values attached to each board square for scoring.
  * @param initPieces Initial pieces on board and in trays.
  * @param playerId Unique id of player.
  * @param startTime Time game began.
  * @param endTime Time game ended.
  */
case class GameBase(
  gameId: ID,
  gameParams: GameParams,
  initPieces: InitPieces,
  pointValues: List[List[Int]],
  playerId: ID,
  startTime: Instant,
  endTime: Option[Instant]
) {
  def scorer = Scorer(gameParams.dimension, gameParams.trayCapacity, pointValues)

  def end: GameBase = {
    val now = Instant.now()
    this.copy(endTime = Some(now))
  }
}

object GameBase {
  def apply(
    gameParams: GameParams,
    initPieces: InitPieces,
    pointValues: List[List[Int]],
    playerId: ID,
  ): GameBase = {
    val now = Instant.now()
    val languageCode = gameParams.languageCode
    val realLanguageCode = if (!languageCode.isEmpty) languageCode else english
    val params = gameParams.copy(languageCode = realLanguageCode)
    GameBase(stringId(), params, initPieces, pointValues, playerId, now, None)
  }

}

File: haskell-server/src/BoardGame/Server/Domain/GameBase.hs
--
-- Copyright 2017-2018 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE FlexibleContexts #-}

module BoardGame.Server.Domain.GameBase (
    GameBase(..)
  , mkInitialBase
)
where

import qualified Data.ByteString.Lazy.Char8 as BC
import Data.Time (UTCTime, getCurrentTime)
import GHC.Generics
import Data.Aeson (FromJSON, ToJSON)
import qualified Data.Aeson as Aeson
import Control.Monad.IO.Class (MonadIO(..))
import Control.Monad.Except (MonadError(..))

import qualified Bolour.Util.MiscUtil as Util (mkUuid)
import BoardGame.Common.Domain.PieceProviderType (PieceProviderType)
import BoardGame.Common.Domain.GameParams (GameParams, GameParams(GameParams))
import qualified BoardGame.Common.Domain.GameParams as GameParams
import BoardGame.Common.Domain.Piece (Piece)
import BoardGame.Common.Domain.InitPieces (InitPieces, InitPieces(InitPieces))
import qualified BoardGame.Common.Domain.InitPieces as InitPieces
import BoardGame.Common.Domain.InitPieces (InitPieces)
import qualified BoardGame.Common.Domain.InitPieces as InitPieces
import qualified BoardGame.Server.Domain.Player as Player
import BoardGame.Server.Domain.GameError (GameError)
import BoardGame.Server.Domain.Player (Player, Player(Player))
import BoardGame.Server.Domain.Tray (Tray)
import qualified BoardGame.Server.Domain.Tray as Tray
import BoardGame.Server.Domain.PieceProvider (PieceProvider)
import qualified BoardGame.Server.Domain.PieceProvider as PieceProvider

-- Note. Excludes initial board to save space.
-- Initial board is reconstructed by using InitPiece {piecePoints}.

-- | The base state of the game - properties that are conceptually
--   immutable once set, and can be used to reconstitute the initial state
--   of the game.
data GameBase = GameBase {
    gameId :: String
  , gameParams :: GameParams
  , pointValues :: [[Int]]
  , initPieces :: InitPieces
  , initTrays :: [Tray]
  -- , playerName :: String
  , playerId :: String
  , startTime :: UTCTime
  , endTime :: Maybe UTCTime
}
  deriving (Eq, Show, Generic)

instance FromJSON GameBase
instance ToJSON GameBase

mkInitialBase :: (MonadError GameError m, MonadIO m) =>
  GameParams -> [[Int]] -> InitPieces -> Player -> PieceProvider -> m (GameBase, PieceProvider)

mkInitialBase gameParams pointValues initPieces player pieceProvider = do
  let GameParams { dimension, trayCapacity, languageCode} = gameParams
  let InitPieces {userPieces, machinePieces} = initPieces
  let Player {playerId, name = playerName} = player
  gameId <- Util.mkUuid
  startTime <- liftIO getCurrentTime
  (userTray, pieceProvider') <- Tray.mkTray pieceProvider trayCapacity userPieces
  (machineTray, pieceProvider'') <- Tray.mkTray pieceProvider' trayCapacity machinePieces
  let trays = [userTray, machineTray]
      endTime = Nothing
      base = GameBase gameId gameParams pointValues initPieces trays
             playerId startTime endTime
  return (base, pieceProvider'')


